sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/NGrams.scala

```diff
@@ -0,0 +1,258 @@
+           accuracyExpression: Expression) = {
+    this(child, nExpression, kExpression, NGrams.getAccuracy(kExpression, accuracyExpression), 0, 0)
+  }
+
+  def this(child: Expression, nExpression: Expression, kExpression: Expression) = {
+    this(child, nExpression, kExpression, Literal(0))
+  }
+
+  private lazy val n: Int = nExpression.eval().asInstanceOf[Int]
+  private lazy val k: Int = kExpression.eval().asInstanceOf[Int]
+  private lazy val accuracy: Int = accuracyExpression.eval().asInstanceOf[Int]
+
+  override def inputTypes: Seq[AbstractDataType] = {
+    Seq(TypeCollection(ArrayType(StringType, false), ArrayType(ArrayType(StringType, false))),
+      IntegerType, IntegerType, IntegerType)
+  }
+
+  val isArrayOfString = child.dataType == ArrayType(StringType, false) ||
+    child.dataType == ArrayType(StringType, true)
+
+  override def checkInputDataTypes(): TypeCheckResult = {
+    val defaultCheck = super.checkInputDataTypes()
+    if (defaultCheck.isFailure) {
+      defaultCheck
+    } else if (!nExpression.foldable || !kExpression.foldable || !accuracyExpression.foldable) {
+      TypeCheckFailure(s"The accuracy or percentage provided must be a constant literal")
+    } else if (accuracy <= 0) {
+      TypeCheckFailure(
+        s"The accuracy provided must be a positive integer literal (current value = $accuracy)")
+    }
+    else {
+      TypeCheckSuccess
+    }
+  }
+
+  override def createAggregationBuffer(): NGramBuffer = {
+    new NGramBuffer(n, k, accuracy, new HashMap[Vector[UTF8String], Double]())
+  }
+
+  def updateArray(genericArrayData: GenericArrayData, buffer: NGramBuffer, inputRow: InternalRow) {
+    val values = (0 until genericArrayData.numElements()).map(genericArrayData.get(_, StringType).
+      asInstanceOf[UTF8String]).toVector
+    val nGrams = getNGrams(values, n)
+    nGrams.foreach(buffer.add(_))
+    buffer.trim()
+  }
+  override def update(buffer: NGramBuffer, inputRow: InternalRow): NGramBuffer = {
+    if (isArrayOfString) {
+      updateArray(child.eval(inputRow).asInstanceOf[GenericArrayData], buffer, inputRow)
+    }
+    else {
+      val arrayOfArray = child.eval(inputRow).asInstanceOf[GenericArrayData]
+      for (i <- 0 until arrayOfArray.numElements()) {
+        updateArray(arrayOfArray.getArray(i).asInstanceOf[GenericArrayData], buffer, inputRow)
+      }
+    }
+    buffer
+  }
+
+  override def merge(buffer: NGramBuffer, input: NGramBuffer): NGramBuffer = {
+    buffer.merge(input)
+    buffer.trim()
+    buffer
+  }
+
+  override def eval(buffer: NGramBuffer): Any = {
+    val topKNGrams = buffer.getTopKNGrams().map((keyValuePair: (Vector[UTF8String], Double)) => {
+      val arrayKey = new GenericArrayData(Vector(new GenericArrayData(keyValuePair._1)))
+      val arrayValue = new GenericArrayData(Vector(keyValuePair._2))
+      new ArrayBasedMapData(arrayKey, arrayValue)
+    }).toVector
+   new GenericArrayData(topKNGrams)
```

Comment:
wrong ident. There are many style issues. Can you run style check locally and fix them?